File: backend/app/services/notifications.py
from sqlalchemy import func
from sqlalchemy.orm import Session

from app.models.notification import Notification


def create_notification(
    db: Session,
    user_id: int,
    type: str,
    title: str,
    message: str,
    related_alert_id: int | None = None,
) -> Notification:
    notification = Notification(
        user_id=user_id,
        type=type,
        title=title,
        message=message,
        related_alert_id=related_alert_id,
    )
    db.add(notification)
    db.commit()
    db.refresh(notification)
    return notification


def list_notifications(
    db: Session,
    user_id: int,
    limit: int = 20,
    offset: int = 0,
) -> list[Notification]:
    return (
        db.query(Notification)
        .filter(Notification.user_id == user_id)
        .order_by(Notification.created_at.desc())
        .offset(offset)
        .limit(limit)
        .all()
    )


def mark_notification_as_read(
    db: Session,
    user_id: int,
    notification_id: int,
) -> bool:
    notification = (
        db.query(Notification)
        .filter(
            Notification.id == notification_id,
            Notification.user_id == user_id,
        )
        .first()
    )
    if not notification:
        return False
    notification.is_read = True
    db.commit()
    return True


def mark_all_notifications_as_read(db: Session, user_id: int) -> int:
    count = (
        db.query(Notification)
        .filter(
            Notification.user_id == user_id,
            Notification.is_read.is_(False),
        )
        .update({"is_read": True})
    )
    db.commit()
    return count


def get_unread_count(db: Session, user_id: int) -> int:
    return (
        db.query(func.count(Notification.id))
        .filter(
            Notification.user_id == user_id,
            Notification.is_read.is_(False),
        )
        .scalar()
    )

File: frontend/src/lib/notifications.ts
const API_BASE = process.env.NEXT_PUBLIC_API_BASE_URL ?? "http://localhost:8000";

export interface Notification {
  id: number;
  user_id: number;
  type: string;
  title: string;
  message: string;
  is_read: boolean;
  related_alert_id: number | null;
  created_at: string;
}

function authHeaders(token: string): Record<string, string> {
  return {
    Authorization: `Bearer ${token}`,
    "Content-Type": "application/json",
  };
}

export async function fetchNotifications(
  token: string,
  limit = 20,
  offset = 0,
): Promise<Notification[]> {
  const res = await fetch(
    `${API_BASE}/api/notifications?limit=${limit}&offset=${offset}`,
    { headers: authHeaders(token) },
  );
  if (!res.ok) throw new Error(`API error: ${res.status}`);
  return res.json() as Promise<Notification[]>;
}

export async function fetchUnreadCount(token: string): Promise<number> {
  const res = await fetch(`${API_BASE}/api/notifications/unread-count`, {
    headers: authHeaders(token),
  });
  if (!res.ok) throw new Error(`API error: ${res.status}`);
  const data = (await res.json()) as { unread_count: number };
  return data.unread_count;
}

export async function markNotificationAsRead(
  token: string,
  id: number,
): Promise<void> {
  const res = await fetch(`${API_BASE}/api/notifications/${id}/read`, {
    method: "POST",
    headers: authHeaders(token),
  });
  if (!res.ok) {
    const err = (await res.json()) as { detail?: string };
    throw new Error(err.detail ?? `Failed to mark as read (${res.status})`);
  }
}

export async function markAllNotificationsAsRead(
  token: string,
): Promise<void> {
  const res = await fetch(`${API_BASE}/api/notifications/read-all`, {
    method: "POST",
    headers: authHeaders(token),
  });
  if (!res.ok) {
    const err = (await res.json()) as { detail?: string };
    throw new Error(
      err.detail ?? `Failed to mark all as read (${res.status})`,
    );
  }
}
